compiler/src/iree/compiler/Codegen/TransformDialectStrategies/GPU/MatmulTensorCoreStrategy.cpp

```diff
@@ -235,7 +235,16 @@ void iree_compiler::gpu::buildMatmulTensorCoreStrategy(
   // Step 3. Hoist the padding of the output operand above the reduction loop.
   // The resulting fillOp will be mapped with the contraction using an SIMD
   // programming model.
-  Value fillOpH = buildHoistOutputPaddingOp(b, variantH, paddedMatmulOpH);
+  Value fillOpH;
+  if (!strategy.alignedRes()) {
+    fillOpH = buildHoistOutputPaddingOp(b, variantH, paddedMatmulOpH);
+  } else {
+    fillOpH = b.create<transform::MatchOp>(variantH,
+                                           linalg::FillOp::getOperationName());
+    ApplyPatternsOpPatterns config;
```

Comment:
Can you add a comment as to why this is necessary?
I'll try and remove it in a followup.

Reply:
(Leaving a longer comment here first)

First for rematching the fill, I see that I can just use the fill handle above from block distribution so that part is unnecessary and I will drop it with the rebase.

The important part here is running canonicalization. For aligned right now we are relying on canonicalization to simplify the `tensor.pad` such that once we rewrite in DPS we get a `linalg.copy`. For example, without canonicalization we will have something like:
```
%padded = tensor.pad %extracted_slice_8 nofold low[%c0_12, %c0_12] high[%14, %15] {
^bb0(%arg5: index, %arg6: index):
   tensor.yield %cst_11 : f16
} : tensor<128x16xf16> to tensor<128x16xf16>
```
and after DPS
```
%17 = linalg.fill ins(%cst_16 : f16) outs(%16 : tensor<128x16xf16>) -> tensor<128x16xf16>
%inserted_slice = tensor.insert_slice %extracted_slice_8 into %17[%c0_12, %c0_12] [128, 16] [1, 1] : tensor<128x16xf16> into tensor<128x16xf16>
```
Whereas if we do canonicalization between the pad and DPS then we get
```
%padded = tensor.pad %extracted_slice_2 nofold low[%c0, %c0] high[%c0, %c0] {
^bb0(%arg5: index, %arg6: index):
  tensor.yield %cst : f16
} : tensor<128x16xf16> to tensor<128x16xf16>
```
into
```
%11 = bufferization.alloc_tensor() : tensor<128x16xf16>
%12 = linalg.copy ins(%extracted_slice_2 : tensor<128x16xf16>) outs(%11 : tensor<128x16xf16>) -> tensor<128x16xf16>
```
(also annoyingly if we do canonicalization after rewriting to dps instead of before all the copies/pads get folded away)
```
%extracted_slice_2 = tensor.extract_slice %extracted_slice[0, %arg3] [128, 16] [1, 1] : tensor<128x1024xf16> to tensor<128x16xf16>
%extracted_slice_3 = tensor.extract_slice %extracted_slice_0[%arg3, 0] [16, 128] [1, 1] : tensor<1024x128xf16> to tensor<16x128xf16>
%11 = linalg.matmul ins(%extracted_slice_2, %extracted_slice_3 : tensor<128x16xf16>, tensor<16x128xf16>) outs(%arg4 : tensor<128x128xf16>) -> tensor<128x128xf16>
```

Right now, `HoistOutputPaddingOp` runs canonicalization as a last step, but we shouldn't do hoisting if the result is unaligned. Also IIRC if we canonicalize before hoisting then there is a chance that the extract_slice on the loop variable gets folded out causing hoisting to fail (definitely some brittleness to address there) so canonicalizing after the pad isn't an option. To summarize

1. If any pads are aligned then we must canonicalize before converting to DPS to get the copy
2. If the output is unaligned, we will run canonicalization anyway so aligned input pads will get the canonicalization they need anyway.
3. Canonicalization cannot run between padding and hoisting and we do not want to hoist an aligned output.

I can just make canonicalization unconditionally run after conditionally doing the output hoisting to remove the `else`, although I figured it would be ugly to run canonicalization back-to-back in the unaligned case. WDYT?